### PR TITLE
Adding tests for AiredEpisodeOrderComparer.

### PR DESCRIPTION
**Pull request by ankenyr**
**Changes**
Creates tests with >80% code coverage for AiredEpisodeOrderComparer.

**Issues**
First step in addressing https://github.com/jellyfin/jellyfin/issues/6269

### CODE REVIEW

**Comment by crobibero** [comment]
I know the namespace that you added tests for is "Emby.Server.Implementations", but please just add it to the existing "Jellyfin.Server.Implementations.Tests" project

**Comment by ankenyr** [comment]
Ok renamed and also moved the files.

**Comment by crobibero** [comment]
I think removing the existing files was missed in the PR update

**Comment by ankenyr** [comment]
This should be good to pull in, let me know fi I am missing anything more.

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            if (err == true)
+            {
+                Assert.Throws<ArgumentNullException>(() => cmp.Compare(x, y));
+            }
+            else
+            {
+                Assert.Equal(expected, cmp.Compare(x, y));
+                if (expected == 1)
+                {
+                    Assert.Equal(expected * -1, cmp.Compare(y, x));
```
Comment (crobibero) [inline tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:28]:
```suggestion
                    Assert.Equal(-expected, cmp.Compare(y, x));
```

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+using MediaBrowser.Controller.Entities.TV;
+using Xunit;
+
+namespace Jellyfin.Server.Implementations.Tests.Sorting
+{
+    public class AiredEpisodeOrderComparerTests
+    {
+        [Theory]
+        [ClassData(typeof(EpisodeTestData))]
+        public void Test1(BaseItem x, BaseItem y, int expected, bool err)
```
Comment (crobibero) [inline tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:16]: I'm not sure what a good test name would be here, but I don't think "Test1" is descriptive enough

Comment (ankenyr) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs]: That is embarrassing, updated it to be named a bit better.

**Review by crobibero — APPROVED** [review]

**Review by cvium — CHANGES_REQUESTED** [review]
The test does too many things. It should be split into at least two tests where the focus is on "failures" and "successes" respectively.

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            var cmp = new AiredEpisodeOrderComparer();
+            if (err == true)
+            {
+                Assert.Throws<ArgumentNullException>(() => cmp.Compare(x, y));
+            }
+            else
+            {
+                Assert.Equal(expected, cmp.Compare(x, y));
+                if (expected == 1)
```
Comment (cvium) [inline tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:26]: Why the if? Isn't it always the case that `Compare(x, y) == -cmp.Compare(y, x)`?

Comment (ankenyr) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:26]: If the result is 0 then switching x and y provides no added benefit. I added this to not waste time checking something that is not required. I can remove it if desired.

**Comment by ankenyr** [comment]
Separated out the good cases from the errors.

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs (hunk 3 above)
Comment (ankenyr) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs]: Sorry I should not reply until after I drink my coffee in the morning.
Yes it should always be the case that `Compare(x, y) == -cmp.Compare(y, x)`
My goal here was to have as much code coverage as possible. Inside of compare there is logic to determine if x or y is a episode or special.

https://github.com/jellyfin/jellyfin/blob/804614315819c8c17c576b2cdfb683177b59eb7a/Emby.Server.Implementations/Sorting/AiredEpisodeOrderComparer.cs#L77-L82

If someone forgot to write the code on lines 77 to 80, then the tests may not catch that however if we only do compare(x,y) and do not try the reverse.

I did not try the reverse of 0 cases where 0 was returned because it felt unnecessary.

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+using MediaBrowser.Controller.Entities.TV;
+using Xunit;
+
+namespace Jellyfin.Server.Implementations.Tests.Sorting
+{
+    public class AiredEpisodeOrderComparerTests
+    {
+        [Theory]
+        [ClassData(typeof(EpisodeBadData))]
+        public void AiredEpisodeOrderCompareErrorTest(BaseItem x, BaseItem y)
```
Comment (cvium) [inline tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:16]: Names should be very self-explanatory.
```suggestion
        public void Compare_GivenNull_ThrowsArgumentNullException(BaseItem x, BaseItem y)
```

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            var cmp = new AiredEpisodeOrderComparer();
+            Assert.Throws<ArgumentNullException>(() => cmp.Compare(x, y));
+        }
+
+        [Theory]
+        [ClassData(typeof(EpisodeTestData))]
+        public void AiredEpisodeOrderCompareTest(BaseItem x, BaseItem y, int expected)
```
Comment (cvium) [inline tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:24]:
```suggestion
        public void Compare_GivenValidBaseItems_Success(BaseItem x, BaseItem y, int expected)
```

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs (hunk 3 above)
Comment (Bond-009) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs]: Don't see the point of the if statement, just always do the check, you won't notice a difference in test execution speed

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                Assert.Equal(-expected, cmp.Compare(y, x));
+            }
+        }
+
+        private class EpisodeBadData : IEnumerable<object?[]>
+        {
+            public IEnumerator<object?[]> GetEnumerator()
+            {
+                yield return new object?[] { null, new Episode() };
+                yield return new object?[] { new Episode() };
```
Comment (Bond-009) [inline tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:40]:
```suggestion
                yield return new object?[] { new Episode(), null };
```

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+                yield return new object?[] { new Episode { ParentIndexNumber = 3, IndexNumber = 1 }, new Episode { ParentIndexNumber = 0, IndexNumber = 1, AirsAfterSeasonNumber = 1, AirsBeforeEpisodeNumber = 2 }, 1 };
+
+                yield return new object?[] { new Episode { ParentIndexNumber = 1, IndexNumber = 1 }, new Episode { ParentIndexNumber = 0, IndexNumber = 1, AirsBeforeSeasonNumber = 1 }, 1 };
+                yield return new object?[] { new Episode { ParentIndexNumber = 1, IndexNumber = 2 }, new Episode { ParentIndexNumber = 0, IndexNumber = 1, AirsBeforeSeasonNumber = 1, AirsBeforeEpisodeNumber = 2 }, 1 };
+                yield return new object?[] { new Episode { ParentIndexNumber = 1 }, new Episode { ParentIndexNumber = 0, IndexNumber = 1, AirsBeforeSeasonNumber = 1, AirsBeforeEpisodeNumber = 2 }, 0 };
+                yield return new object?[] { new Episode { ParentIndexNumber = 1, IndexNumber = 3 }, new Episode { ParentIndexNumber = 0, IndexNumber = 1, AirsBeforeSeasonNumber = 1, AirsBeforeEpisodeNumber = 2 }, 1 };
```
Comment (Bond-009) [inline tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:77]: Please format this properly, putting it all on one line is hard to read IMO.

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs (hunk 7 above)
Comment (ankenyr) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:77]: Done I think. I separated the episodes and result onto separate lines. I did not separate each of the episodes attributes onto lines as it felt like it would make it too long. I can do that though if you all wish it.

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs (hunk 5 above)
Comment (ankenyr) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs:24]: commited

**Comment by Bond-009** [comment]
https://github.com/jellyfin/jellyfin/pull/6403#discussion_r689123639

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs (hunk 3 above)
Comment (ankenyr) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs]: Sorry guess I missed this one, not very used to the github PR review process and find it hard to follow comments.

**Review by Bond-009 — APPROVED** [review]

File: tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs (hunk 3 above)
Comment (cvium) [reply tests/Jellyfin.Server.Implementations.Tests/Sorting/AiredEpisodeOrderComparerTests.cs]: There's a fairly new feature on GitHub that might help you
image

**Review by cvium — APPROVED** [review]
Thank you for enduring the PR process 😄
